server/src/main/java/org/elasticsearch/search/aggregations/pipeline/BucketSortPipelineAggregator.java

```diff
@@ -95,22 +93,22 @@ public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext
             return originalAgg.create(new ArrayList<>(buckets.subList(from, Math.min(from + currentSize, bucketsCount))));
         }
 
-        int queueSize = Math.min(from + currentSize, bucketsCount);
-        PriorityQueue<ComparableBucket> ordered = new TopNPriorityQueue(queueSize);
+        List<ComparableBucket> ordered = new ArrayList<>();
         for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
             ComparableBucket comparableBucket = new ComparableBucket(originalAgg, bucket);
             if (comparableBucket.skip() == false) {
-                ordered.insertWithOverflow(new ComparableBucket(originalAgg, bucket));
+                ordered.add(new ComparableBucket(originalAgg, bucket));
             }
         }
 
-        int resultSize = Math.max(ordered.size() - from, 0);
+        Collections.sort(ordered, Collections.reverseOrder());
```

Comment:
The fact we have to sort on reverse order seems a bit awkward here. What if we simplify? I think we can now change `ComparableBucket.compareTo` to ignore the sort order and behave like a `compareTo` promises to behave. Then in this line, we can choose to reverse order or not depending on the sort order. I think this will make it much cleaner. What do you think?

Reply:
Agreed. That would make it easier to see what's happening.

Reply:
On second thought, we can't ignore the sort order in `ComparableBucket.compareTo` since we may be sorting buckets on multiple fields (each with a different order). Still, we can make the method return what we would normally expect from a `compareTo`. Then we won't have to sort the list in reverse and the code will make more sense overall. I will push the new version soon.

Also, the current `compareTo` (and the tests) consider `null` to be greater than any other value. Out of curiosity, is this some standard assumption, at least when it comes to Elasticsearch?

Reply:
Thanks for this and sorry for the delayed response. Holidays kicked in :-) Happy new year!

Elasticsearch is handling `null` in ordinary search/sort by putting those values at the end (or the beginning when the order is descending). It also lets the user change that behaviour via the `missing` parameter. Having said that, I would argue it is overkill to support this in the context of the `bucket_sort` aggregation.

Looking more into the code, it actually looks like `ComparableBucket` could never really get any `null` values. There are 3 scenarios:

- the sort field is `_key`; keys will be non-null I believe
- `gap_policy` is `skip`; the bucket won't exist at all
- `gap_policy` is `insert_zeros`; the value of the bucket will be `0.0` and will be sorted accordingly.

So, it most probably is dead code.